### PR TITLE
Update Saving games documentation for Godot 4.0

### PR DESCRIPTION
**Pull request by xdl**
Updating saving games doc to Godot4 (including a best effort attempt at C# with dotnet6). Overview of changes:

* File → FileAcces methods
  * Removing call to `.close()` in GDScript and using `using` in C#
  * `get_len` → `get_length`
* `res.instance()` to `res.instantiate()`
* `node.filename` → `node.scene_file_path`
* (C#) JSON method update (Print → Stringify)
* (C#) Minor clarification on setting lowercase 'position' - seems to have tripped up people in the past in 3.x too: https://godotengine.org/qa/136145/c%23-problem-with-save-load-node-positions-set-to-0-0
* (GDScript) Adding the omitted `var json_string = save_game.get_line()`
* (C#) object → Variant when declaring a dictionary
  * Was done to get around the compile error: `The generic type argument must be a Variant compatible type: 'object'` - not sure if this is the correct way to go about it, if someone can confirm that'd be great

### CODE REVIEW

File: tutorials/io/saving_games.rst [hunk 1]
```diff
@@ -273,18 +266,18 @@ load function:
...
         // it represents.
-        saveGame.Open("user://savegame.save", (int)File.ModeFlags.Read);
+        using var saveGame = FileAccess.Open("user://savegame.save", FileAccess.ModeFlags.Read);
 
-        while (saveGame.GetPosition() < saveGame.GetLen())
+        while (saveGame.GetPosition() < saveGame.GetLength())
         {
             // Get the saved dictionary from the next line in the save file
-            var nodeData = new Godot.Collections.Dictionary<string, object>((Godot.Collections.Dictionary)JSON.Parse(saveGame.GetLine()).Result);
+            var nodeData = new Godot.Collections.Dictionary<string, Variant>((Godot.Collections.Dictionary)JSON.Parse(saveGame.GetLine()));
```
[the remark this answers is not in the document]
Comment (xdl) [reply tutorials/io/saving_games.rst]: Good spot! Fair point about error handling; ParseString seems to return an empty object when it's malformed - any suggestions on how to handle? Not sure how to pull up docs for this method.

File: tutorials/io/saving_games.rst [hunk 2]
```diff
@@ -92,7 +92,7 @@ The save function will look like this:
     {
         return new Godot.Collections.Dictionary<string, object>()
```
[the remark this answers is not in the document]
Comment (xdl) [reply tutorials/io/saving_games.rst]: Good spot!

File: tutorials/io/saving_games.rst [hunk 3]
```diff
@@ -168,14 +166,13 @@ way to pull the data out of the file as well.
...
+        using var saveGame = FileAccess.Open("user://savegame.save", FileAccess.ModeFlags.Write);
 
         var saveNodes = GetTree().GetNodesInGroup("Persist");
         foreach (Node saveNode in saveNodes)
         {
             // Check the node is an instanced scene so it can be instanced again during load.
-            if (saveNode.Filename.Empty())
+            if (String.IsNullOrEmpty(saveNode.SceneFilePath))
             {
                 GD.Print(String.Format("persistent node '{0}' is not an instanced scene, skipped", saveNode.Name));
```
[the remark this answers is not in the document]
Comment (xdl) [reply tutorials/io/saving_games.rst]: Sounds good; updated the other usage too.

File: tutorials/io/saving_games.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (raulsntos) [reply tutorials/io/saving_games.rst]: We should probably copy the GDScript version and create an instance of `JSON` so we can use the instance method `JSON.Parse` that returns an `Error`, then you can check if it's `Error.Ok`.

The documentation is the same as in GDScript, you can find it in the [class reference](https://docs.godotengine.org/en/latest/classes/class_json.html), the only difference is the names are converted from snake_case to PascalCase in C#, everything else is the same.

[the remark this answers is not in the document]
Comment (xdl) [reply tutorials/io/saving_games.rst]: Cool thanks - have just updated that.

File: tutorials/io/saving_games.rst [hunk 4]
```diff
@@ -273,30 +266,37 @@ load function:
...
         // it represents.
-        saveGame.Open("user://savegame.save", (int)File.ModeFlags.Read);
+        using var saveGame = FileAccess.Open("user://savegame.save", FileAccess.ModeFlags.Read);
 
-        while (saveGame.GetPosition() < saveGame.GetLen())
+        while (saveGame.GetPosition() < saveGame.GetLength())
         {
+            var jsonString = saveGame.GetLine();
+            // Creates the helper class to interact with JSON
+            var json = new JSON();
```
Comment (raulsntos) [inline tutorials/io/saving_games.rst:275]: nit: use the same whitespace as in the GDScript example.

```suggestion
            var jsonString = saveGame.GetLine();

            // Creates the helper class to interact with JSON
            var json = new JSON();

```

File: tutorials/io/saving_games.rst [hunk 5]
```diff
@@ -273,30 +266,37 @@ load function:
...
+            var parseResult = json.Parse(jsonString);
+            if (parseResult != Error.Ok) {
+                GD.Print($"JSON Parse Error: {json.GetErrorMessage()} in {jsonString} at line {json.GetErrorLine()}");
+                continue;
+            }
```
Comment (raulsntos) [inline tutorials/io/saving_games.rst:280]: C# style is to put the opening brackets in a new line.

```suggestion
            if (parseResult != Error.Ok)
            {
                GD.Print($"JSON Parse Error: {json.GetErrorMessage()} in {jsonString} at line {json.GetErrorLine()}");
                continue;
            }
```

File: tutorials/io/saving_games.rst [hunk 6]
```diff
@@ -273,30 +266,37 @@ load function:
...
+
             // Get the saved dictionary from the next line in the save file
```
Comment (raulsntos) [inline tutorials/io/saving_games.rst:282]: Update the comment to follow the GDScript example.

```suggestion
            // Get the data from the JSON object
```

File: tutorials/io/saving_games.rst [hunk 7]
```diff
@@ -273,30 +266,37 @@ load function:
...
             var newObject = (Node)newObjectScene.Instantiate();
             GetNode(nodeData["Parent"].ToString()).AddChild(newObject);
-            newObject.Set("Position", new Vector2((float)nodeData["PosX"], (float)nodeData["PosY"]));
+            newObject.Set("position", new Vector2((float)nodeData["PosX"], (float)nodeData["PosY"]));
 
             // Now we set the remaining variables.
-            foreach (KeyValuePair<string, object> entry in nodeData)
+            foreach (KeyValuePair<string, Variant> entry in nodeData)
             {
                 string key = entry.Key.ToString();
```
Comment (raulsntos) [inline tutorials/io/saving_games.rst:294]: `entry.Key` should already be a string.

```suggestion
                string key = entry.Key;
```

File: tutorials/io/saving_games.rst [hunk 8]
```diff
@@ -168,34 +166,31 @@ way to pull the data out of the file as well.
...
+                GD.Print($"persistent node '{saveNode.Name}' is missing a Save() function, skipped");
                 continue;
             }
 
             // Call the node's save function.
             var nodeData = saveNode.Call("Save");
 
             // Store the save dictionary as a new line in the save file.
-            saveGame.StoreLine(JSON.Print(nodeData));
+            saveGame.StoreLine(JSON.Stringify(nodeData));
```
Comment (raulsntos) [inline tutorials/io/saving_games.rst:192]: In the GDScript example, the `JSON.Stringify` invocation is in a separate line and contains an useful comment about what it does.

**Review by raulsntos — COMMENTED** [review]
Thanks for the contribution! Looks pretty good already, just a few nitpicks.

**Comment by xdl** [comment]
Cheers for the feedback and the suggestions - have added those in!

File: tutorials/io/saving_games.rst [hunk 9]
```diff
@@ -168,34 +166,34 @@ way to pull the data out of the file as well.
...
             {
-                GD.Print(String.Format("persistent node '{0}' is missing a Save() function, skipped", saveNode.Name));
+                GD.Print($"persistent node '{saveNode.Name}' is missing a Save() function, skipped");
                 continue;
             }
 
             // Call the node's save function.
             var nodeData = saveNode.Call("Save");
 
+            # JSON provides a static method to serialized JSON string
```
Comment (raulsntos) [inline tutorials/io/saving_games.rst:191]: Comments in C# use `//`.

```suggestion
            // JSON provides a static method to serialized JSON string
```

Comment (xdl) [reply tutorials/io/saving_games.rst]: Doh good spot

**Review by raulsntos — APPROVED** [review]
LGTM

**Review by Calinou — APPROVED** [review]
Thanks!

**Pull request merged**